File: src/lib/backtest/execution.ts
/**
 * Execution Model
 *
 * Models transaction costs, slippage, and order execution for backtesting.
 */

// ============================================================================
// Types
// ============================================================================

export interface ExecutionCosts {
  /** Commission as a percentage (e.g., 0.001 = 0.1%) */
  commissionPct: number;
  /** Slippage in basis points (e.g., 5 = 0.05%) */
  slippageBps: number;
}

export interface ExecutionResult {
  /** Actual fill price after costs */
  fillPrice: number;
  /** Total cost as percentage of trade value */
  totalCostPct: number;
  /** Commission cost */
  commissionCost: number;
  /** Slippage cost */
  slippageCost: number;
}

// ============================================================================
// Default Costs
// ============================================================================

// Binance Futures taker fees
export const BINANCE_FUTURES_COSTS: ExecutionCosts = {
  commissionPct: 0.0004, // 0.04% taker fee
  slippageBps: 3,        // ~3 bps slippage for liquid pairs
};

// More conservative estimates
export const CONSERVATIVE_COSTS: ExecutionCosts = {
  commissionPct: 0.001,  // 0.1% total fees
  slippageBps: 10,       // 10 bps slippage
};

// ============================================================================
// Slippage Models
// ============================================================================

/**
 * Fixed slippage model
 * Simple percentage-based slippage
 */
export function fixedSlippage(price: number, slippageBps: number, isBuy: boolean): number {
  const slippagePct = slippageBps / 10000;
  if (isBuy) {
    return price * (1 + slippagePct);
  } else {
    return price * (1 - slippagePct);
  }
}

/**
 * Volume-adjusted slippage model
 * Higher slippage for larger orders relative to volume
 *
 * @param price - Order price
 * @param orderSize - Order size in base currency
 * @param avgVolume - Average volume for the asset
 * @param baseSlippageBps - Base slippage in bps
 * @param isBuy - Whether this is a buy order
 */
export function volumeAdjustedSlippage(
  price: number,
  orderSize: number,
  avgVolume: number,
  baseSlippageBps: number,
  isBuy: boolean
): number {
  // Market impact scales with sqrt of order size relative to volume
  const orderPctOfVolume = orderSize / avgVolume;
  const impactMultiplier = 1 + Math.sqrt(orderPctOfVolume * 100);
  const adjustedSlippageBps = baseSlippageBps * impactMultiplier;

  return fixedSlippage(price, adjustedSlippageBps, isBuy);
}

/**
 * Spread-based slippage model
 * Uses bid-ask spread for more accurate execution
 */
export function spreadBasedSlippage(
  midPrice: number,
  spreadPct: number,
  isBuy: boolean
): number {
  const halfSpread = spreadPct / 2;
  if (isBuy) {
    return midPrice * (1 + halfSpread);
  } else {
    return midPrice * (1 - halfSpread);
  }
}

// ============================================================================
// Execution Simulator
// ============================================================================

/**
 * Simulate order execution with costs
 */
export function simulateExecution(
  price: number,
  costs: ExecutionCosts,
  isBuy: boolean
): ExecutionResult {
  // Apply slippage
  const priceAfterSlippage = fixedSlippage(price, costs.slippageBps, isBuy);

  // Calculate costs
  const slippageCost = Math.abs(priceAfterSlippage - price) / price;
  const commissionCost = costs.commissionPct;
  const totalCostPct = slippageCost + commissionCost;

  return {
    fillPrice: priceAfterSlippage,
    totalCostPct,
    commissionCost,
    slippageCost,
  };
}

/**
 * Calculate round-trip costs (entry + exit)
 */
export function calculateRoundTripCosts(costs: ExecutionCosts): number {
  // Commission on both entry and exit
  const commissions = costs.commissionPct * 2;

  // Slippage on both entry and exit
  const slippage = (costs.slippageBps / 10000) * 2;

  return commissions + slippage;
}

/**
 * Apply costs to trade PnL
 */
export function applyTradeCosts(
  grossPnlPct: number,
  costs: ExecutionCosts
): number {
  const roundTripCosts = calculateRoundTripCosts(costs);
  return grossPnlPct - roundTripCosts;
}

// ============================================================================
// Break-even Analysis
// ============================================================================

/**
 * Calculate minimum trade profit needed to break even after costs
 */
export function calculateBreakEvenPnl(costs: ExecutionCosts): number {
  return calculateRoundTripCosts(costs);
}

/**
 * Calculate required win rate for profitability given costs
 *
 * @param avgWinPct - Average winning trade percentage
 * @param avgLossPct - Average losing trade percentage (positive number)
 * @param costs - Execution costs
 */
export function calculateRequiredWinRate(
  avgWinPct: number,
  avgLossPct: number,
  costs: ExecutionCosts
): number {
  const roundTripCosts = calculateRoundTripCosts(costs);

  // Net win/loss after costs
  const netWin = avgWinPct - roundTripCosts;
  const netLoss = avgLossPct + roundTripCosts;

  // Required win rate for break-even: netLoss / (netWin + netLoss)
  return netLoss / (netWin + netLoss);
}

// ============================================================================
// Pair Trading Execution
// ============================================================================

/**
 * Simulate pair trade execution (enter both legs simultaneously)
 */
export function simulatePairEntry(
  price1: number,
  price2: number,
  hedgeRatio: number,
  costs: ExecutionCosts,
  isLongSpread: boolean
): {
  fillPrice1: number;
  fillPrice2: number;
  totalCostsPct: number;
} {
  // Long spread = long asset1, short asset2
  // Short spread = short asset1, long asset2
  const leg1IsBuy = isLongSpread;
  const leg2IsBuy = !isLongSpread;

  const exec1 = simulateExecution(price1, costs, leg1IsBuy);
  const exec2 = simulateExecution(price2, costs, leg2IsBuy);

  // Total costs weighted by position sizes
  // Assuming notional is 1 for asset1 and hedgeRatio for asset2
  const weight1 = 1 / (1 + hedgeRatio);
  const weight2 = hedgeRatio / (1 + hedgeRatio);

  const totalCostsPct = exec1.totalCostPct * weight1 + exec2.totalCostPct * weight2;

  return {
    fillPrice1: exec1.fillPrice,
    fillPrice2: exec2.fillPrice,
    totalCostsPct,
  };
}

/**
 * Simulate pair trade exit
 */
export function simulatePairExit(
  price1: number,
  price2: number,
  hedgeRatio: number,
  costs: ExecutionCosts,
  wasLongSpread: boolean
): {
  fillPrice1: number;
  fillPrice2: number;
  totalCostsPct: number;
} {
  // Exit is opposite of entry
  return simulatePairEntry(price1, price2, hedgeRatio, costs, !wasLongSpread);
}

/**
 * Calculate pair trade PnL
 *
 * IMPORTANT: PnL is normalized by position weights to avoid amplification effects.
 * Each leg contributes proportionally to total capital at risk.
 *
 * For example, if hedgeRatio = 5.0:
 * - Asset1 weight: 1/(1+5) = 16.7%
 * - Asset2 weight: 5/(1+5) = 83.3%
 *
 * This prevents small price differences from causing massive PnL swings.
 */
export function calculatePairTradePnl(
  entryPrice1: number,
  entryPrice2: number,
  exitPrice1: number,
  exitPrice2: number,
  hedgeRatio: number,
  isLongSpread: boolean
): number {
  const return1 = (exitPrice1 - entryPrice1) / entryPrice1;
  const return2 = (exitPrice2 - entryPrice2) / entryPrice2;

  // Normalize by position weights (beta-neutral with equal risk contribution)
  const absHedgeRatio = Math.abs(hedgeRatio);
  const weight1 = 1 / (1 + absHedgeRatio);
  const weight2 = absHedgeRatio / (1 + absHedgeRatio);

  // Long spread: profit when spread widens (asset1 outperforms asset2)
  // Short spread: profit when spread narrows (asset2 outperforms asset1)
  if (isLongSpread) {
    return weight1 * return1 - weight2 * return2;
  } else {
    return weight2 * return2 - weight1 * return1;
  }
}

File: python/src/backtest/execution.py
"""
Execution Model

Models transaction costs, slippage, and order execution for backtesting.
"""

from dataclasses import dataclass
from typing import Tuple
import numpy as np


@dataclass
class ExecutionCosts:
    """Execution cost parameters"""

    commission_pct: float  # Commission as percentage (e.g., 0.001 = 0.1%)
    slippage_bps: float  # Slippage in basis points (e.g., 5 = 0.05%)


# Default costs for Binance Futures
BINANCE_FUTURES_COSTS = ExecutionCosts(
    commission_pct=0.0004,  # 0.04% taker fee
    slippage_bps=3.0,  # ~3 bps slippage for liquid pairs
)

# Conservative estimates
CONSERVATIVE_COSTS = ExecutionCosts(
    commission_pct=0.001,  # 0.1% total fees
    slippage_bps=10.0,  # 10 bps slippage
)


def calculate_round_trip_costs(costs: ExecutionCosts) -> float:
    """
    Calculate round-trip costs (entry + exit)

    Returns:
        Total round-trip cost as percentage
    """
    # Commission on both entry and exit
    commissions = costs.commission_pct * 2

    # Slippage on both entry and exit
    slippage = (costs.slippage_bps / 10000) * 2

    return commissions + slippage


def calculate_pair_trade_pnl(
    entry_price1: float,
    entry_price2: float,
    exit_price1: float,
    exit_price2: float,
    hedge_ratio: float,
    is_long_spread: bool,
) -> float:
    """
    Calculate pair trade PnL

    IMPORTANT: PnL is normalized by position weights to avoid amplification effects.
    Each leg contributes proportionally to total capital at risk.

    For example, if hedgeRatio = 5.0:
    - Asset1 weight: 1/(1+5) = 16.7%
    - Asset2 weight: 5/(1+5) = 83.3%

    This prevents small price differences from causing massive PnL swings.

    Args:
        entry_price1: Entry price for asset 1
        entry_price2: Entry price for asset 2
        exit_price1: Exit price for asset 1
        exit_price2: Exit price for asset 2
        hedge_ratio: β from cointegrating regression
        is_long_spread: True if long spread (long asset1, short asset2)

    Returns:
        PnL as percentage
    """
    # Calculate returns for each leg
    return1 = (exit_price1 - entry_price1) / entry_price1
    return2 = (exit_price2 - entry_price2) / entry_price2

    # Normalize by position weights (beta-neutral with equal risk contribution)
    abs_hedge_ratio = abs(hedge_ratio)
    weight1 = 1 / (1 + abs_hedge_ratio)
    weight2 = abs_hedge_ratio / (1 + abs_hedge_ratio)

    # Long spread: profit when spread widens (asset1 outperforms asset2)
    # Short spread: profit when spread narrows (asset2 outperforms asset1)
    if is_long_spread:
        return weight1 * return1 - weight2 * return2
    else:
        return weight2 * return2 - weight1 * return1


def apply_trade_costs(gross_pnl_pct: float, costs: ExecutionCosts) -> float:
    """
    Apply execution costs to gross PnL

    Args:
        gross_pnl_pct: Gross PnL as percentage
        costs: Execution costs

    Returns:
        Net PnL after costs
    """
    round_trip_costs = calculate_round_trip_costs(costs)
    return gross_pnl_pct - round_trip_costs


def calculate_break_even_pnl(costs: ExecutionCosts) -> float:
    """
    Calculate minimum trade profit needed to break even after costs

    Returns:
        Break-even PnL percentage
    """
    return calculate_round_trip_costs(costs)
